mlflow/gateway/providers/mosaicml.py

```diff
@@ -0,0 +1,236 @@
+from typing import Any, Dict, List
+
+from fastapi import HTTPException
+from fastapi.encoders import jsonable_encoder
+
+from mlflow.exceptions import MlflowException
+from mlflow.gateway.config import MosaicMLConfig, RouteConfig
+from mlflow.gateway.providers.base import BaseProvider
+from mlflow.gateway.providers.utils import rename_payload_keys, send_request
+from mlflow.gateway.schemas import chat, completions, embeddings
+
+
+class MosaicMLProvider(BaseProvider):
+    def __init__(self, config: RouteConfig) -> None:
+        super().__init__(config)
+        if config.model.config is None or not isinstance(config.model.config, MosaicMLConfig):
+            raise TypeError(f"Unexpected config type {config.model.config}")
+        self.mosaicml_config: MosaicMLConfig = config.model.config
+
+    async def _request(self, model: str, payload: Dict[str, Any]) -> Dict[str, Any]:
+        headers = {"Authorization": f"{self.mosaicml_config.mosaicml_api_key}"}
+        return await send_request(
+            headers=headers,
+            base_url=self.mosaicml_config.mosaicml_api_base
+            or "https://models.hosted-on.mosaicml.hosting",
+            path=model + "/v1/predict",
+            payload=payload,
+        )
+
+    def _parse_chat_messages_to_prompt(self, messages: List[chat.RequestMessage]) -> str:
+        """
+        This parser is based on the format described in
+        https://huggingface.co/blog/llama2#how-to-prompt-llama-2 .
+        The expected format is:
+            "<s>[INST] <<SYS>>
+            {{ system_prompt }}
+            <</SYS>>
+
+            {{ user_msg_1 }} [/INST] {{ model_answer_1 }} </s><s>[INST] {{ user_msg_2 }} [/INST]"
+        """
+        if any(m1.role == m2.role for m1, m2 in zip(messages, messages[1:])):
+            raise MlflowException.invalid_parameter_value(
+                "Consecutive messages cannot have the same 'role'.",
+            )
+        prompt = ""
+        for m in messages:
+            role = m.role
+            content = m.content
+            if role == "system":
+                prompt += f"<s>[INST] <<SYS>> {content} <</SYS>>"
+            elif role == "user":
+                inst = f"{content} [/INST]"
+                if prompt.endswith("<</SYS>>"):
```

Comment:
Update this,  please:  it should be  `if not prompt.endswith("<</SYS>>"):`

Can we manually QA this before merging? 

Make sure the cases of user instructions with or without system prompt work fine. Thanks!!

Reply:
(If feasible, adding a unit test for these cases is much better!)

Reply:
Added the case!